### PR TITLE
[dtensor] add a implicit replication flag

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115566
* #115565
* #115564
* __->__ #115297

This PR adds a experimental implicit replication support for DTensor to
inter-op with torch.Tensor, basically under this context manager DTensor
could work together with torch.Tensor by assuming the torch.Tensor
sharding layout is replicated.

Note that this is risky for DTensor so we don't turn it on by default,
but for certain cases where it is for sure replicated, user can use this
to allow DTensor and Tensor computation work together

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc

### CODE REVIEW

**Review by awgu — APPROVED** [review]
Thanks for adding this!

File: test/distributed/_tensor/test_dtensor.py [hunk 1]
```diff
@@ -742,6 +742,18 @@ def test_redistribute_sub_mesh(self):
...
+    def test_implicit_replication(self):
+        mesh = init_device_mesh(self.device_type, (self.world_size,))
+        local_tensor1 = torch.ones(4, 3)
+        sharded_dtensor = DTensor.from_local(local_tensor1, mesh, [Shard(0)])
+
+        from torch.distributed._tensor.experimental import implicit_replication
+
+        with implicit_replication():
+            out_dt = sharded_dtensor + torch.randn(3, device=self.device_type)
+            self.assertEqual(out_dt.placements, [Shard(0)])
```
Comment (awgu) [inline test/distributed/_tensor/test_dtensor.py:755]: Should we include any numeric test to check the result of the add?

What is the expected semantics of adding a replicated `(3,)` local tensor to a `(4, 3)` local tensor? It should broadcast over dim-1?

Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:755]: sure!

> What is the expected semantics of adding a replicated (3,) local tensor to a (4, 3) local tensor? It should broadcast over dim-1?

Yep it would be implicitly broadcasted according on the "4" dim according to the tensor broadcasting rule https://pytorch.org/docs/stable/notes/broadcasting.html#broadcasting-semantics
